internal/addrs/moveable.go

```diff
@@ -16,14 +16,27 @@ type AbsMoveable interface {
 	String() string
 }
 
-// The following are all of the possible AbsMovable address types:
+// The following are all of the possible AbsMoveable address types:
 var (
 	_ AbsMoveable = AbsResource{}
 	_ AbsMoveable = AbsResourceInstance{}
 	_ AbsMoveable = ModuleInstance(nil)
 	_ AbsMoveable = AbsModuleCall{}
 )
 
+// AbsMoveableResource is an AbsMoveable that is either a resource or a resource
+// instance.
+type AbsMoveableResource interface {
+	AbsMoveable
+	Type() string
```

Comment:
I like the idea of this interface, but worry that `Type() string` might be too generic a signature because it seems likely to inadvertently overlap with other methods we might add on non-resource things that have types in the future.

Initially I was going to suggest `ResourceType() string` as a way to make it a bit more specific, but then I realized that what all implementers of this interface ought to have in common is having an `addrs.AbsResource`, and so maybe the following signature would address the "too general" concern while also giving us a more convenient interface to some other resource-ish things too:

```go
type AbsMoveableResource interface {
    AbsMoveable
    AffectedResource() AbsResource
}
```

Your `absFrom.Type()` elsewhere would then be `absFrom.AffectedResource().Type`, which is a little more verbose but I think still similarly convenient (no more individual type assertions for each) and maybe subjectively more explicit to a reader of the call.

What do you think? :thinking:

Reply:
Happy with this. I agree that `Type()` is a bit too generic.